docs/fobnail-api.md

```diff
@@ -0,0 +1,459 @@
+connected. To check connection status, the server issues CoAP Ping requests,
+which clients **must** respond to in order to avoid losing their objects.
+
+When an object is created, the server responds with **2.01**, `Location-Path`
+specifying the resource's path (a numeric object ID). Object IDs are specific to
+a client - a client cannot access object IDs of other clients. `Location-Path`
+is a string option, whereas object ID is an integer, so it must be encoded as a
+string. Object IDs may be passed as part of the CBOR payload (as an integer) or
+in URI (as a string).
+
+If the `Content-Format` option is not present, Fobnail Token assumes that
+`Content-Format` is `application/octet-stream`. All requests which pass
+CBOR-encoded data **must** set `Content-Format` option to `application/cbor` for
+Fobnail Token to interpret the payload as CBOR. The server responds with
+**4.00** (Bad Request) if unknown or invalid' Content-Format' is given. If
+passing CBOR-encoded data which is improperly encoded or not valid (e.g.,
+required fields missing) server responds with **4.00**. Passing
+`application/cbor` to an endpoint that expects `application/octet-stream` is
+invalid and will result in a **4.00** response.
+
+The client may send one or more `Accept` options which **must** be handled by
+the server. If the server can not respond in the format requested by the
+`Accept` options, it **must** respond with **4.06** (Not Acceptable).
+
+[Conditional Request Options](https://datatracker.ietf.org/doc/html/rfc7252#section-5.10.8)
+are **not** supported and will result in **4.02** (Bad Option). The payload
+should contain a human-readable message describing what options are not
+supported. See
+[RFC7252 section 5.4.1](https://datatracker.ietf.org/doc/html/rfc7252#section-5.4.1)
+for details.
+
+Each response from the Fobnail Token **must** contain the `Content-Format`
+option unless it is an error response with a diagnostic payload. See
+[RFC7252](https://datatracker.ietf.org/doc/html/rfc7252#section-5.5.1) for
+details.
+
+### Error responses
+
+Fobnail Token may return error codes as defined by the CoAP specification. An
+error response **may** contain an additional context sent as a payload. All
+cacheable error responses **must** have `Max-Age` set to 0.
+
+Error response must not contain the `Content-Format` option. Message (if present)
+must be passed as a raw UTF-8 string.
+
+### Fobnail Data format
+
+Fobnail uses CBOR (with a few exceptions) to transmit structured data from
+Fobnail to Attester and from Attester to Fobnail. We use a subset of
+[RFC7049](https://datatracker.ietf.org/doc/html/rfc7049) (no tagging and no
+extensions).
+
+Clients are **required** to set `Content-Format` to `Application/CBOR` when
+transmitting CBOR-encoded data.
+
+### Data signing
+
+A signature is generated by taking an already CBOR-encoded object, a nonce
+(generated by Fobnail and sent as part of the request), and signing the hash of
```

Comment:
Fobnail doesn't generate requests anymore. Also, since this is not part of single request/response pair, nonce would probably have to be returned in next request from client, in case another client (or attacker) tries to connect in between those two requests (generate nonce and send signed data).

Reply:
@krystian-hebel We can't trust that the nonce we get from client is real. If attacker can replay signed message he may also provide replayed nonce. In some cases we could use nonce client-provided nonce - when Fobnail signs data and client verifies it but not when Fobnail verifies that data.

Reply:
I'm not talking about generating nonce on platform, just not stripping it from next request. That way server doesn't have to check it against every generated nonce, in case there are multiple simultaneously open connections in the future.

Also, how attester is supposed to get nonce for `/attest`?

Reply:
Server keeps a single nonce for each client so it has to compare only with that nonce. I added /nonce endpoint for reading nonce.